src/libxively/event_dispatcher/xi_time_event.c

```diff
@@ -0,0 +1,426 @@
+    }
+
+    /* exit with out_state value */
+    return out_state;
+
+err_handling:
+    return out_state;
+}
+
+xi_time_event_t* xi_time_event_get_top( xi_vector_t* vector )
+{
+    /* PRE-CONDITIONS */
+    assert( NULL != vector );
+
+    if ( 0 == vector->elem_no )
+    {
+        return NULL;
+    }
+
+    xi_time_event_t* top_one = ( xi_time_event_t* )vector->array[0].selector_t.ptr_value;
+
+    /* the trick is to bubble the element to the end of the vector and then to delete it*/
+    if ( vector->elem_no > 1 )
+    {
+        xi_time_event_move_to_the_end( vector, 0 );
+        xi_vector_del( vector, vector->elem_no - 1 );
+    }
+    else
+    {
+        xi_vector_del( vector, vector->elem_no - 1 );
+    }
+
+    xi_time_event_dispose_time_event( top_one );
+
+    return top_one;
+}
+
+xi_time_event_t* xi_time_event_peek_top( xi_vector_t* vector )
+{
+    /* PRE-CONDITIONS */
+    assert( NULL != vector );
+
+    if ( 0 == vector->elem_no )
+    {
+        return NULL;
+    }
+
+    xi_time_event_t* top_one = ( xi_time_event_t* )vector->array[0].selector_t.ptr_value;
+
+    return top_one;
+}
+
+xi_state_t xi_time_event_restart( xi_vector_t* vector,
+                                  xi_time_event_handle_t* time_event_handle,
+                                  xi_time_t new_time )
+{
+    /* PRE-CONDITIONS */
+    assert( NULL != vector );
+    assert( NULL != time_event_handle );
+
+    /* the element can be found with O(1) complexity cause we've been updating each
+     * element's position during every operation that could've broken it */
+
+    xi_vector_index_type_t index = *time_event_handle->position;
+
+    /* check for the correctness of the time_event_handle */
+    if ( index >= vector->elem_no || index < 0 )
+    {
+        return XI_ELEMENT_NOT_FOUND;
+    }
+
+    /* let's update the key of this element */
+    xi_time_event_t* time_event =
+        ( xi_time_event_t* )vector->array[index].selector_t.ptr_value;
+
+    /* sanity check on the time handle */
+    assert( time_event->time_event_handle == time_event_handle );
+
+    time_event->time_of_execution = new_time;
+
+    xi_time_event_bubble_and_sort_down(
+        vector, xi_time_event_bubble_and_sort_up( vector, index ) );
+
+    return XI_STATE_OK;
+}
+
+xi_state_t xi_time_event_cancel( xi_vector_t* vector,
+                                 xi_time_event_handle_t* time_event_handle,
+                                 xi_time_event_t** cancelled_time_event )
+{
+    /* PRE-CONDITIONS */
+    assert( NULL != vector );
+    assert( NULL != time_event_handle );
+    assert( NULL != time_event_handle->position );
+    assert( NULL != cancelled_time_event );
+
+    /* the element we would like to remove should be at position described by the
+     * time_event_handle */
+
+    xi_vector_index_type_t index = *time_event_handle->position;
+
+    if ( index >= vector->elem_no || index < 0 )
+    {
+        return XI_ELEMENT_NOT_FOUND;
```

Comment:
Assertions are used as an argument type validation for debug for situations that must not occur - the error handling is used for situations that may happen during the runtime.